src/ListMessageSplitter.ts

```diff
@@ -0,0 +1,319 @@
+                    rest.items.unshift(sized.items.pop()!)
+
+                    if (sized.items.length > 0) {
+                        return {sized, rest};
+                    } else {
+                        // If this was the first item, `sized` is empty, return null.
+                        return {sized: null, rest}
+                    }
+                }
+
+                current_item = rest.items.shift();
+            }
+
+            // ??? We somehow did not go over the size limit when testing it per item?
+            // This is a weird state, as we should have already caught this with the
+            // `this.size() <= desired_size` conditional.
+            LogService.warn("ListMessageSplitter", `encountered end of while loop, required max ${desired_size} size, current item is ${sized.size()}`)
+
+            return {sized, rest: null}
+        }
+    }
+
+    public size(): number {
+        let with_header = this.header !== null ? this.header.size() + "\n".length + "<br>".length : 0;
+
+        return with_header
+            + (this.items.length > 0 ? UL_TAG_WRAP_SIZE : 0)
+            + this.items.reduce((prev, curr, _idx, _arr) => prev + curr.size(), 0);
+    }
+
+    // Render this listing into a messagesnippet.
+    public render(): MessageSnippet {
+        let current: MessageSnippet = {
+            html: "",
+            text: ""
+        };
+
+        for (let item of this.items) {
+            current.html += `<li>${item.html}</li>`
+            current.text += ` * ${item.text}\n`
+        }
+
+        current.html = this.items.length > 0 ? `<ul>${current.html}</ul>` : "";
+
+        if (this.header !== null) {
+            current.html = this.header.html + "<br>".length + current.html;
+            current.text = this.header.text + "\n" + current.text;
+        }
+
+        return current;
+    }
+}
+
+// A class that allows splitting items and headers into multiple messages.
+export class ListMessageSplitter {
+    private items: (MessageListItem | MessageListHeader)[] = [];
+
+    constructor() {
+    }
+
+    public add_header(html: string, text: string) {
+        this.add(new MessageListHeader(html, text))
+    }
+
+    public add_paragraph(html: string, text: string) {
+        this.add(new MessageListItem(html, text))
+    }
+
+    public add(paragraph: MessageListItem | MessageListHeader) {
+        this.items.push(paragraph)
+    }
+
+    // Convert this.items into listings.
+    private get_listings(): MessageListing[] {
+        let current_listings: MessageListing[] = [];
+        let current_listing = new MessageListing(null);
+
+        for (let item of this.items) {
+            if (item instanceof MessageListHeader) {
+                if (current_listing.header !== null) {
+                    current_listings.push(current_listing);
+
+                    current_listing = new MessageListing(item);
+                } else {
+                    current_listing.header = item;
```

Comment:
If I'm understanding this right, this check probably should be done as they add the header so that the state is always consistent (if there is a header in `this.items` then `current_listing.header` will be set to that), not when using a utility method.

Reply:
Ok, I think I have misread the code.

Though, I do think that assigning the new listing a header after instantiation of the listing (rather than when it is instantiated) is a problem/smell.
Especially when we're saying that a listing can have a header, but it doesn't have to be the first item in a listing, that is making the code unnecessarily complicated. 

I don't think `add`, `add_paragraph`, `add_header` etc should push to `this.items` in this case, it would be better if they would produce listings as they were invoked and then behaviour like "one header per listing" becomes pretty easy to express and validate (it becomes implicit).

Though this does mean if you have a message that is a bunch of paragraphs with a header part way through, you would create a new listing rather than make that the header of existing paragraphs. Though that seems to make sense anyways?